enhancements/update/eus-upgrades-mvp.md

```diff
@@ -0,0 +1,368 @@
+- We will reduce workload disruption associated with pod rescheduling during rolling
+reboots
+
+
+### Non-Goals
+
+- This enhancement does not attempt to remove *any* steps along the serial
+  upgrade path from 4.6 to 4.7 to 4.8 to 4.9 to 4.10 *including* reboots. A
+  follow-up enhancement will address those concerns.
+
+## Proposal
+
+Before we dive into user stories it's useful that you're familiar with CVO and
+Console feature which are relevant to how upgrades and channel options are
+presented to the admin. The CVO and Console as of 4.6 are now aware of other
+channels containing the cluster's current version. This means that if 4.6.50 is
+a non EUS version and 4.6.51 is an EUS version both the CVO and Console can
+display a different set of channel options. This means that 4.6.50 would see
+candidate-4.6, fast-4.6, stable-4.6, eus-4.6, candidate-4.7, fast-4.7, and
+stable-4.7 channels.
+
+Where as 4.6.51 would see only eus-4.6 and eus-4.10 assuming we have enabled EUS
+4.6 to 4.10 upgrades. After switching to eus-4.10 the cluster would offer
+upgrades to 4.7 versions present in the eus-4.10 channel, when that completes
+the cluster would offer 4.8 versions present in the eus-4.10 channel, repeat
+until the cluster has completed its upgrade to 4.10.
+
+### User Stories
+
+#### APIServer - Notify cluster admins of API use which will be removed between 4.7 and 4.10
+
+We have pre-existing mechanisms to alert admins to their use of APIs which are
+slated for removal in a future release. Currently those mechanisms do not look
+forward far enough to span 4.6 to 4.10, we will need to add the ability to
+backport that API metadata to 4.6 EUS versions once we have a firm understanding
+of all removals between 4.6 and 4.10.
+
+### MCO - Enforce OpenShift's defined host component version skew policies
+
+The MCO, will set Upgradeable=False whenever any MachineConfigPool has one more
+more nodes present which falloutside of a define list of constraints. For
+instance, if OpenShift has a defined Kubelet Version Skew of N-1, the node
+constraints enforced by the MCO defined in OCP 4.7 could be as follows:
+
+node.status.nodeInfo.kubeletVersion:
+ - v1.20
+ - v1.19
+
+If the policy were to change to allow for a version skew of N-2, v1.18 would be
+added to the list of acceptable matches. The MCO is not responsible for defining
+these constrains and constraints are only widened whenever we have CI testing
+proves them to be safe.
+
+These changes will need to be backported to 4.7 prior to 4.7 EOL.
+
+#### OLM - Allow Operators to define inclusive compatible version ranges
+
+We should allow maintainers to define an inclusive version range for their Operator.
+
+#### OLM - Enforce defined OLM managed Operator compatibility
+
+If a managed Operator defines a maxKubeVersion or maxOCPVersion which would be
+violated by upgrading to the next minor OLM must set Upgradeable=False and
+enumerate the constraint in the condition message.
```

Comment:
this, and a few other guards in the enhancement, are vulnerable to update-retarget vulnerabilities that openshift/cluster-version-operator#460 has been in flight for a while trying to address.  For example, today you could:

1. Be running 4.6.z.
2. Request an update to 4.7.z'.
3. CVO begins updating to 4.7.z'.
4. CVO requests recommended updates from 4.7.z', and hears about 4.8.z".
5. User accepts requested update to 4.8.z" before the 4.7.z' OLM operator had come out to check it's children's max versions against 4.8 and set `Upgradeable=False`.
6. Cluster core hits 4.8.z" and some OLM operators fail on compat violations.

This isn't a blocker for this effort; we've wanted to address the CVO-retarget issue for a while.  But it's been slow, low-priority work, and we probably want to get it going again and drive it to a conclusion as a blocker for updating out of eus-4.6.  I dunno if we have a structured way to represent blockers like that; maybe a whiteboard keyword?

Reply:
Lets call that out as a story here and make sure that we're giving that the priority and attention necessary.

Reply:
new story here https://github.com/openshift/enhancements/pull/585/files#diff-91c1b36fc770bc12ef9147ff77360aa00fb1f79d3df986a471d75d1c4a33891cR153-R170